From: Esben Haabendal <esben@geanix.com>
Date: Thu, 11 Apr 2024 14:19:23 +0200
Subject: serial: imx: Introduce timeout when waiting on transmitter empty
Git-commit: e533e4c62e9993e62e947ae9bbec34e4c7ae81c2
Patch-mainline: 6.10-rc1
References: CVE-2024-40967 bsc#1227891

By waiting at most 1 second for USR2_TXDC to be set, we avoid a potential
deadlock.

In case of the timeout, there is not much we can do, so we simply ignore
the transmitter state and optimistically try to continue.

[js] read_poll_timeout_atomic is not available, use readx_poll_timeout_atomic

Signed-off-by: Esben Haabendal <esben@geanix.com>
Acked-by: Marc Kleine-Budde <mkl@pengutronix.de>
Link: https://lore.kernel.org/r/919647898c337a46604edcabaf13d42d80c0915d.1712837613.git.esben@geanix.com
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 drivers/tty/serial/imx.c |    8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

--- a/drivers/tty/serial/imx.c
+++ b/drivers/tty/serial/imx.c
@@ -38,6 +38,7 @@
 #include <linux/of.h>
 #include <linux/of_device.h>
 #include <linux/io.h>
+#include <linux/iopoll.h>
 #include <linux/dma-mapping.h>
 
 #include <asm/irq.h>
@@ -1805,7 +1806,7 @@ imx_console_write(struct console *co, co
 {
 	struct imx_port *sport = imx_ports[co->index];
 	struct imx_port_ucrs old_ucr;
-	unsigned int ucr1;
+	unsigned int ucr1, usr2;
 	unsigned long flags = 0;
 	int locked = 1;
 
@@ -1837,8 +1838,9 @@ imx_console_write(struct console *co, co
 	 *	Finally, wait for transmitter to become empty
 	 *	and restore UCR1/2/3
 	 */
-	while (!(readl(sport->port.membase + USR2) & USR2_TXDC));
-
+	readx_poll_timeout_atomic(readl, sport->port.membase + USR2,
+				  usr2, usr2 & USR2_TXDC,
+				  0, USEC_PER_SEC);
 	imx_port_ucrs_restore(&sport->port, &old_ucr);
 
 	if (locked)
